c-s CLI: columns and population options

muzarski opened the pull request:
# Changes
- Refactored the `Param` trait, by extracting most of the common logic to the wrapping type
- introduced `DistributionParam`
- implemented parsing logic for `-col` option
- implemented parsing logic for `-pop` option

piodul commented on src/bin/cql-stress-cassandra-stress/settings/param/distribution.rs:
```diff
@@ -0,0 +1,153 @@
+/// Distribution-specific arguments parser.
+trait DistributionParser: Sync {
+    fn parse(&self, distribution_args: &[&str]) -> Result<Box<dyn DistributionFactory>>;
+    fn specification_and_description(&self) -> (&'static str, &'static str);
```
Shouldn't this be two methods?

piodul commented on src/bin/cql-stress-cassandra-stress/settings/param/distribution.rs:
```diff
@@ -0,0 +1,153 @@
+        let map = HashMap::new();
+        map
+    };
+}
```
I think this is overengineered. Why not have a function that does pattern matching one the string and returns a distribution parser?

muzarski replied:
Yeah, I can see your point. I only did it for the "efficiency" purposes so we don't have to return a `Box`'ed `DistributionParser` with each invocation of such function. I wanted to allocate each parser once.

But as you said, this might be overengineered, since there wouldn't be many allocations of `DistributionParser` during parsing anyway.

piodul replied:
AFAIK `Box<T>` where `T` is a zero-sized type does not allocate.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/param/distribution.rs:
```diff
@@ -0,0 +1,153 @@
+use super::Param;
+
+/// Distribution-specific arguments parser.
+trait DistributionParser: Sync {
```
We already have `src/distribution.rs`, can we reuse it? Cassandra-stress and scylla-bench use similar syntax for distributions, so I wrote that code keeping in mind that it might be possible to use it for the c-s frontend.

muzarski commented:
v2:
- rebased on top of master
- removed utility functions decomposing the CLI argument into distribution description payload - using functionalities implemented in `src/distribution.rs` instead
- removed `DistributionParser` trait (with its implementations)
- moved specific distribution parsing logic from `DistributionParser` implementations to specific `DistributionFactories`.
- moved description printing from `DistributionParser` implementations to distribution factories
- added two new commits (since runtime logic has already been merged):
  - using user-provided distribution (`-pop dist` parameter) as distribution which `RowGenerator` samples `pk_seeds` from
  - using `-col` option parameters in the runtime, including:
    - adjusted table creation/insertion queries based on the columns provided by the user ('-col n' or '-col names' parameters)
    - using cell-size distribution (`RowGenerator` samples the size of the blob data that is going to be generated from this distribution) based on user-provided distribution (`-col size` parameter)

muzarski commented:
v3:
- removed unnecessary TODO next to `RowGeneratorFactory::settings` field.

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/sequence.rs:
```diff
@@ -55,6 +56,58 @@ impl Distribution for SeqDistribution {
+
+impl SeqDistributionFactory {
+    pub fn parse_from_description(desc: Description<'_>) -> Result<Box<dyn DistributionFactory>> {
+        let build_err_msg = || {
+            format!(
+                "Invalid parameter list for sequence distribution: {:?}",
+                desc.args
+            )
+        };
```
This pattern looks a bit tedious - you have to add `with_context` before all question marks - and it's easy to forget about it and introduce a question mark without adding context.

I suggest introducing an inner function like suggested here: https://users.rust-lang.org/t/attach-context-to-all-propagated-errors-with-anyhow/45020/2 so that you only need to take care of attaching context in a single place.

piodul replied:
This applies to other commits that use this pattern as well.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/param/parser.rs:
```diff
@@ -147,6 +148,21 @@ impl ParamsParser {
+
+        self.params.push(Rc::clone(&param) as ParamCell);
+        DistributionParamHandle::new(param)
+    }
```
Why can't we use an existing `simple_param`/`simple_subparam`/`multi_param` for the distribution?

muzarski replied:
`simple_param` parsing logic could be reused for sure. The only difference is help-printing logic.
Look at the part of output of `help -insert`:
```
$ ./cassandra-stress help -insert

Usage: -insert [revisit=DIST(?)] [visits=DIST(?)] partitions=DIST(?) [batchtype=?] select-ratio=DIST(?) row-population-ratio=DIST(?)

  [revisit=DIST(?)]: The distribution with which we revisit partial writes (see visits); implicitly defines size of revisit collection
      EXP(min..max)                        An exponential distribution over the range [min..max]
      EXTREME(min..max,shape)              An extreme value (Weibull) distribution over the range [min..max]
      QEXTREME(min..max,shape,quantas)     An extreme value, split into quantas, within which the chance of selection is uniform
      GAUSSIAN(min..max,stdvrng)           A gaussian/normal distribution, where mean=(min+max)/2, and stdev is (mean-min)/stdvrng
      GAUSSIAN(min..max,mean,stdev)        A gaussian/normal distribution, with explicitly defined mean and stdev
      UNIFORM(min..max)                    A uniform distribution over the range [min, max]
      FIXED(val)                           A fixed distribution, always returning the same value
      SEQ(min..max)                        A fixed sequence, returning values in the range min to max sequentially (starting based on seed), wrapping if necessary.
      Preceding the name with ~ will invert the distribution, e.g. ~exp(1..10) will yield 10 most, instead of least, often
      Aliases: extr, qextr, gauss, normal, norm, weibull

... rest of parameters ...
```

In addition to parameters info (name, usage etc.), the distribution parameters list all of the available distributions.

piodul replied:
Would it make sense to adjust `simple_param` so that it optionally supports printing additional help?

muzarski replied:
Yes, I think so.

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/fixed.rs:
```diff
@@ -58,3 +58,9 @@ impl FixedDistributionFactory {
     }
 }
+
+impl std::fmt::Display for FixedDistributionFactory {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "FIXED({})", self.0)
+    }
+}
```
Why do we need to implement Display for factories? The commit message does not explain that.

muzarski replied:
To display the configuration before running the runtime (I'll include it in the commit message).

piodul commented on src/bin/cql-stress-cassandra-stress/settings/cs_args_bad_test.in:
```diff
@@ -15,3 +15,7 @@ cassandra-stress read -schema replication(factor==123)
+cassandra-stress write -col size=FIXED(1..10)
+cassandra-stress write -col size=UNIFORM(50)
+cassandra-stress write -col size=SEQ(1..10,50)
+cassandra-stress write -col names=foo,bar,baz n=3
```
Nit: missing newline at the end

piodul commented on src/bin/cql-stress-cassandra-stress/settings/cs_args_good_test.in:
```diff
@@ -22,3 +22,6 @@ cassandra-stress counter_write no-warmup cl=LOCAL_ONE duration=20m -schema repli
 cassandra-stress counter_write no-warmup cl=ALL duration=20m -schema replication(replication_factor=3) keyspace=keyspace2
+cassandra-stress read cl=QUORUM n=10000 -schema replication(factor=1) -rate threads=10 -col n=10 size=UNIFORM(1..20)
+cassandra-stress read cl=QUORUM n=10000 -schema replication(key=value) -rate threads=10 -col names=foo,bar,baz
+cassandra-stress read cl=QUORUM n=10000 -schema replication(factor=1, key=value) -rate threads=10 -col size=FIXED(50)
```
> c-s option: introduce -col option 
>
> Implemented parsing logic for -col option.
> 
> For reference, see:
> ```
> $ ./cassandra-stress help -col
> Usage: -node [n=?] [size=DIST(?)]
>   OR
> Usage: -node names=? [size=DIST(?)]
> ```

That's great, but what does this option do? The commit message shows some help message, but it looks like it explains how to use the `-node` option...?

piodul commented on src/bin/cql-stress-cassandra-stress/settings/cs_args_bad_test.in:
```diff
@@ -18,4 +18,6 @@ cassandra-stress write -rate keyspace=keyspace2
\ No newline at end of file
+cassandra-stress write -col names=foo,bar,baz n=3
+cassandra-stress write -pop dist=foo
+cassandra-stress write -pop dist=SEQ(1..10,50)
```
Missing newline at the end

piodul commented on src/bin/cql-stress-cassandra-stress/settings/param/mod.rs:
```diff
@@ -30,7 +30,7 @@ pub trait ParamImpl {
         param_name: &'static str,
         description: &'static str,
-        default_value: Option<&'static str>,
+        default_value: &Option<String>,
```
- Using `&Option<T>` rather than `Option<&T>` is more awkward to use; e.g. to `map` over a value of the first type one must always call `as_ref()` first.
- Why pass a reference to an owned type? `Option<&str>` is always more flexible than `Option<&String>` or `&Option<String>`. You can always convert a `&String` to `&str` but not the other way around.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/param/simple_param.rs:
```diff
@@ -66,6 +69,9 @@ impl<T: Parsable> ParamImpl for SimpleParam<T> {
         }
         println!("{:<40} {}", usage, description);
+        if let Some(additional_description) = &self.additional_desc {
+            println!("{:<40}", additional_description)
```
What is `{:<40}` needed for? AFAIK it aligns the string to the left, but is there a need to do it?

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/fixed.rs:
```diff
@@ -23,3 +27,36 @@ impl Distribution for FixedDistribution {
+
+impl FixedDistributionFactory {
+    pub fn parse_from_description(desc: Description<'_>) -> Result<Box<dyn DistributionFactory>> {
+        let result = || -> Result<Box<dyn DistributionFactory>> {
```
Nit: if `result` were the name of the _result_ of the lambda, then I'd be fine with it, but it's a name for the _lambda_ itself. That's slightly confusing.

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/sequence.rs:
```diff
@@ -55,6 +56,55 @@ impl Distribution for SeqDistribution {
+
+impl SeqDistributionFactory {
+    pub fn parse_from_description(desc: Description<'_>) -> Result<Box<dyn DistributionFactory>> {
+        let result = || -> Result<Box<dyn DistributionFactory>> {
```
Ditto

piodul commented on src/bin/cql-stress-cassandra-stress/java_generate/distribution/uniform.rs:
```diff
@@ -48,6 +51,61 @@ impl Distribution for UniformDistribution {
+
+impl UniformDistributionFactory {
+    pub fn parse_from_description(desc: Description<'_>) -> Result<Box<dyn DistributionFactory>> {
+        let result = || -> Result<Box<dyn DistributionFactory>> {
```
Ditto

piodul commented on src/bin/cql-stress-cassandra-stress/settings/option/column.rs:
```diff
@@ -0,0 +1,126 @@
+    let size_distribution =
+        parser.distribution_param("size=", Some("fixed(34)"), "Cell size distribution", false);
+
+    // $ ./cassandra-stress help -col
+    // Usage: -node [n=?] [size=DIST(?)]
+    //  OR
+    // Usage: -node names=? [size=DIST(?)]
```
You got rid of the weird usage example for `-col` from the commit message, but I also found it here.

piodul commented on src/bin/cql-stress-cassandra-stress/settings/option/column.rs:
```diff
@@ -0,0 +1,126 @@
+    pub fn print_settings(&self) {
+        println!("Column:");
+        println!("  Column names: {:?}", self.columns);
+        println!("  size distribution: {}", self.size_distribution);
```
Nit: `size` -> `Size`

piodul reviewed (commented):
Some small nitpicks left, otherwise LGTM.

muzarski commented:
v5:
fixed review comments, ready to merge
